### PR TITLE
feat(ts-sdk): use lts version in runtime

### PR DESCRIPTION
**Pull request by TomChv**
This PR is an attempt to fixe #7433, if the CI is green, we can merge it, otherwise I'll dig into what breaking change happens with this new version.

### CODE REVIEW

**Comment by TomChv** [comment]
> Since this is an upgrade of versions, I don't think we need to do a breaking change for this - unless there's something specifically very different in this new version.

Yes it's not supposed to be a breaking change but I wanna verify that it doesn't break the TS SDK

**Comment by TomChv** [comment]
@jedevc There's no breaking change with this PR, that's nice to know!

**Comment by Xiot** [comment]
From the node docs:
> Production applications should only use Active LTS or Maintenance LTS releases.

https://nodejs.org/en/about/previous-releases

Node 22 doesn't become active LTS until the end of October
https://github.com/nodejs/release#release-schedule

File: sdk/typescript/runtime/main.go [hunk 1]
```diff
@@ -13,9 +13,10 @@ import (
 
 const (
 	bunVersion  = "1.0.27"
-	nodeVersion = "21.3"
+	// 20.14.0 is the latest LTS version of node as of 2023-10-24, it ends on 2026-04-30
+	nodeVersion = "iron-alpine3.19"
```
Comment (jedevc) [inline sdk/typescript/runtime/main.go:17]: Can we use the `20.14.0-alpine3.19` tag please instead.

**Comment by jedevc** [comment]
Hm, this is now a version *downgrade*. Looking through https://nodejs.org/en/blog/announcements/v21-release-announce, it looks like 21 doesn't bring *that* many user-facing changes.

That said, I'm not a js programmer - but the downgrade would definitely be unexpected. One option would be to delay something like this to v0.12.0 if we're concerned about breaking changes (just because our CI doesn't hit a breaking change doesn't mean it isn't there) cc @helderco

Potentially, we should also allow configuring the node version used for modules by having a field in `dagger.json` - I know we've also discussed this for python/go (though go makes this a *right* pain).

**Comment by TomChv** [comment]
> That said, I'm not a js programmer - but the downgrade would definitely be unexpected. One option would be to delay something like this to v0.12.0 if we're concerned about breaking changes (just because our CI doesn't hit a breaking change doesn't mean it isn't there) cc @helderco

Yeah that's right, maybe the best thing to do is to close this PR and wait for october until we can put v22 when it's the active production release?

**Comment by jedevc** [comment]
I think the downgrade is probably alright, but we shouldn't do it until v0.12.0 if we do.

Or we can upgrade to v22 now (upgrades are find whenever IMO), and just temporarily suffer while it's not an LTS version (I'm not sure what kind of issue this would give us).

**Comment by TomChv** [comment]
> Or we can upgrade to v22 now (upgrades are find whenever IMO), and just temporarily suffer while it's not an LTS version (I'm not sure what kind of issue this would give us).

In the CI everything looked fine but yeah I don't know either, normally node tries to be backward compatible so it should be okay.

I'll keep the Pr on hold for now and take care of that later in the week

**Comment by Xiot** [comment]
Node upgrades / downgrades are always tricky since its not just the user code that you have to be worried about.  Internally we use `20.11`, and one of our devs was setting up a new machine and he had `20.13` installed.  Things broke.

I like the ability to configure which version of node is used, while having a minimum supported version.  Something like this will probably be needed for the `bun` runtime since that is adding new features on a quicker cadence

**Comment by TomChv** [comment]
> Node upgrades / downgrades are always tricky since its not just the user code that you have to be worried about. Internally we use 20.11, and one of our devs was setting up a new machine and he had 20.13 installed. Things broke.

Alright, I'm closing this PR for now, as you said, it's better to allow node configuration which is possible by checking the `engine.node` field of the package.json.

I'll take care of that later, I'll put a comment in the issue as a reminder

**Comment by helderco** [comment]
> Potentially, we should also allow configuring the node version used for modules by having a field in `dagger.json` - I know we've also discussed this for python/go (though go makes this a _right_ pain).

Python supports this via `pyproject.toml`. TypeScript can do the same, just as it's able to select a different runtime (bun), with `package.json`.

**Comment by Xiot** [comment]
You should probably allow the user to configure the package manager as well.  Having that switch out from beneath you (like it recently did) can cause issues as well.

**Comment by TomChv** [comment]
> Python supports this via pyproject.toml. TypeScript can do the same, just as it's able to select a different runtime (bun), with package.json.

Yes but do Python has no breaking change between package manager version? Because that's the case for TypeScript (see https://github.com/dagger/dagger/pull/7518)

**Pull request closed without merging**